### PR TITLE
Spark: Implement copy-on-write UPDATE

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR implements copy-on-write UPDATE in Spark.

### CODE REVIEW

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (aokolnychyi) [inline spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:113]: This part requires a bit of time to understand. The comment above is also important.

In short, we are using the same relation on both sides of the union. Then `DeduplicateRelations` clones one of them and assigns fresh expression IDs so that attributes on both sides are different. During job planning, we plan the job only once and then replace both `DataSourceV2Relation` with two `DataSourceV2ScanRelation`. The scan relations are different objects but they share the SAME `Scan` instance. During dynamic filtering, we assign each scan relation a dynamic filter. The dynamic filter subquery will be executed only once and then shared by two dynamic filters at runtime. Luckily, Spark reuses subqueries with and without AQE. The reason why we have two dynamic filters is because we don't know which side of the union plan will be executed first.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentAlignmentSupport.scala
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentAlignmentSupport.scala:94]: Some comments to explain what is going on here would be really helpful. Maybe also rename this to something that makes it more clear, like `handleCharTypeLimit`.

If you end up renaming this, also consider accepting an Assignment and optionally transforming it, like this:

```scala
    protected def handleCharLimits(assignment: Assignment): Assignment = {
    val rawKeyType = ...
    if (CharVarcharUtils.hasCharVarchar(rawKeyType)) {
      // build the transformed assignment
    } else {
      assignment
    }
  }
```

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentAlignmentSupport.scala
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentAlignmentSupport.scala:241]: +1 for removing `implicit`.

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java:606]: This fails if AQE is enabled?

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java:666]: Looks like this is no longer needed because `testUpdateWithNotInSubquery` is not ignored anymore.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala:96]: If I understand correctly, sometimes a plan gets transformed, from this:

```
ReplaceData
  Update(id#1 IN (1, 5), data#2 = 'foo')
      V2Relation(db.table, [id#1, data#2])
```

To this:
```
ReplaceData
  Update(id#1 IN (1, 5), data#2 = 'foo')
    Union([id#1, data#2])
      V2Relation(db.table, [id#1, data#2])
      V2Relation(db.table, [id#4, data#5])
```

So this is basically creating the dynamic filter for each scan separately and fixing up the attrs, from the update's IDs to the relation's IDs.

Is there a test case for this, or is it something that you ran into in practice?

It took me a while to understand this (assuming that I do) and I think it would be nice if it were more clear what is being replaced. Renaming `attrMap` to `scanRelationAttrs` would probably help so it is clear we're looking up the right attr for the scan relation.

It would also be good to to add a comment that explains why `u.table.output` is always going to match `relation.output`. In fact, I'm not sure that will. What happens when a column in the data is not used to produce the output? Is it pruned?

For example, `UPDATE t SET id = 1 WHERE true` doesn't need to project `id` from the table. If `id` is not projected, then the `attrMap` is incorrect. Is there a reason why that can't happen?

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:60]: In the future, will we also check whether the delta plan is selected? Iceberg can support delta, but choose not to use it because of table settings.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:54]: Where is `u.resolved` implemented? Does that check alignment since `buildUpdateProjection` requires aligned assignments?

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:139]: It wouldn't be too hard to add a projection with one more column and then use the column value to determine whether to replace values.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:87]: Nit: this could be constructed in a private method or passed in instead of `relation`. I think that would help readability.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:116]: If the condition is empty or `true`, then this doesn't need the remaining rows plan or the union. Should we handle that case explicitly?

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:125]: The name `allRowsPlan` is a little misleading since it isn't all rows in the table. That said, I'm not sure what I would rename it to.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala
Comment (rdblue) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala:96]: Okay, after reading through all the cases for rewriting update, this makes a lot more sense. I think it would have helped me reading through this to either direct to the union case in `RewriteUpdateTable` or to explain the rewrite here and give a little more context on when there will be a union.

As for my other comments, I think it would still be good to rename `attrMap` and I'm not sure whether there could be a case where `u.table.output` doesn't match `relation.output`.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java
Comment (rdblue) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:214]: As with the copy-on-write delete case, I think I'd prefer to preserve the existing file boundaries and row order if the table is unpartitioned and unsorted. I would also do that as the default, if neither the write order or the update order is specified for unpartitioned tables.

**Review by rdblue — CHANGES_REQUESTED** [review]
This is looking really close. Thanks, @aokolnychyi!

I only have a few major comments:
* ~~Because the rewrite requires aligned assignments, I think that it needs to only match plans that are aligned. In the optimizer, we can rely on rule order for this, but the analyzer runs in batches so the plan may become resolved after the alignment rule and before the rewrite rule, leading to a case where misaligned plans are rewritten~~
* ~~It shouldn't be too difficult to add the projection to only run the condition once, instead of once per column~~ (See [this discussion](https://github.com/apache/iceberg/pull/3764/files#r772001356))
* ~~I'm not sure it is safe to assume alignment between the output relation and the input relations for dynamic pruning (`u.table.output.zip(relation.output)`)~~
* I'd prefer to preserve files and ordering by default if there is no distribution mode set. I might be wrong, so like the delete command I think it would be helpful to have a table of the proposed behavior

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentAlignmentSupport.scala
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentAlignmentSupport.scala]: I added `AssignmentUtils` with a separate method dedicated to this functionality.

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java:606]: Unfortunately, yes. It contains a distinct aggregation.

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java:666]: Exactly. It is a valid test now.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (rdblue) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:61]: Nevermind, I saw that Iceberg just doesn't return a `SupportsDelta` class.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala]: I've tried to improve the comment to make it clear and also made a note about the rewrite rule. Let me know if its any better.

> For example, UPDATE t SET id = 1 WHERE true doesn't need to project id from the table. If id is not projected, then the attrMap is incorrect.

It was working correctly as we don't support schema pruning in copy-on-write operations. The only case where it is applicable if we have `true` update condition and then we can discard only columns assigned to literals. Seemed like a very narrow use case. I think we can handle it later as we did not support it in the old implementation. If the condition is not trivial, the assigned expression will be `if ... else oldValue` so we won't be able to prune the column.

That being said, I did modify the logic for building the attribute map. It should work even if we add schema pruning.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:61]: We are checking the constructed `RowLevelOperation` we got from `RowLevelOperationBuilder`. At this point, we passed the command and options to the data source and it reported what strategy should be used. In the future, we will just add another method call here.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala]: It previously relied on the order of rules but I made that explicit. That's a good call, I agree.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:138]: Unfortunately, the optimizer would squash the projections and replace the attribute with its expression. We had this limitation in the old implementation too.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala:132]: Should this use the resolver or are we confident that the case will always match?

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:88]: Not sure I got. Could you elaborate, @rdblue?

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala]: I did move this to another place. Let me know if that's what you meant, @rdblue.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala]: Renamed it to `updatedAndRemainingRowsPlan`. Still not entirely happy.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java
Comment (aokolnychyi) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:214]: Let me think.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala:132]: We could use a resolver to be safe but we are sure the names will match as the other attribute is constructed by cloning the original one. I'll update anyway.

**Comment by aokolnychyi** [comment]
@rdblue, it is ready for another round. I'll create a table of proposed distributions and orderings and will think about the default one. That's still an open item.

**Comment by aokolnychyi** [comment]
Oops, looks like I broke something. Let me see.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (rdblue) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:88]: It's minor, but these lines of code are duplicated in a few places. You could have a method that returns the new relation with the original output, metadata attrs, and optional id attrs to simplify these. Not a big deal.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (rdblue) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:138]: If this is what we'd end up with anyway, then let's hold off. I see that this is what we do in 3.1 anyway so no need to change it.

**Review by rdblue — APPROVED** [review]
+1 when tests are passing. I think the major issues I had are now addressed. We don't need to block on the sort order behavior. As with deletes, we can fix that up without blocking the merge support.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:138]: I think we would need a custom node to handle it. I'll look into it after the release.

**Comment by aokolnychyi** [comment]
@rdblue, had to slightly modify `AssignmentUtils`. It turns out our alignment check was too aggressive. We did not catch this before as we did not invoke that code in `RewriteUpdateTable`.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala]: Switched to using a resolver.

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala
Comment (aokolnychyi) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala]: Sounds good. Added a method.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @aokolnychyi! I've merged this to unblock MERGE INTO and we can follow up on the requested sort orders.

**Pull request merged**
